### ISSUE TITLE
T5 batch inference same input data gives different outputs?

### ISSUE THREAD

**Issue by ArvinZhuang** [huggingface/transformers#6786]
# ❓ Questions & Help

## Details

Hi, I am using a T5 model from a tf check point. I have set model.eval(), but every time I forward the exactly same input data, the outputs were always different. here is my code:
```
import torch
from transformers import *

tokenizer = T5Tokenizer.from_pretrained('t5-base')
config = T5Config.from_pretrained('t5-base')
model = T5ForConditionalGeneration.from_pretrained(
    "../model/t5-base/model.ckpt-1004000", from_tf=True, config=config)
model.eval()

def prediction(documents, query):
    querys = [query] * len(documents)
    encoded_encoder_inputs = tokenizer(documents, padding=True, truncation=True, return_tensors="pt")
    encoded_decoder_inputs = tokenizer(querys, padding=True, truncation=True, return_tensors="pt")
    with torch.no_grad():
        outputs = model(input_ids=encoded_encoder_inputs["input_ids"],
                        labels=encoded_decoder_inputs["input_ids"],
                        attention_mask=encoded_encoder_inputs["attention_mask"])
        batch_logits = outputs[1]
        print(batch_logits)


documents = ['my dog is cute!', "so am I."]
query = "who am I?"
prediction(documents, query)
```

here is my twice outputs (run the above code twice):

```
tensor([[[-14.6796,  -6.2236, -13.3517,  ..., -42.7422, -42.7124, -42.8204],
         [-18.5999,  -4.9112, -10.6610,  ..., -40.7506, -40.7313, -40.8373],
         [-17.3894,  -5.3482, -11.4917,  ..., -41.2223, -41.1643, -41.3228],
         [-18.4449,  -5.9145, -12.0056,  ..., -42.3857, -42.3579, -42.4859]],
        [[-15.7967,  -6.9833, -14.5827,  ..., -41.3168, -41.0326, -40.9567],
         [-18.4241,  -5.7193, -12.0748,  ..., -40.1744, -40.0635, -39.9045],
         [-19.5852,  -5.1691, -12.7764,  ..., -42.2655, -42.0788, -41.9885],
         [-20.3673,  -3.6864, -12.5264,  ..., -40.1189, -40.0787, -39.8976]]])
```

```
tensor([[[-14.6796,  -6.2236, -13.3517,  ..., -42.7422, -42.7124, -42.8204],
         [-18.5848,  -4.9116, -10.6607,  ..., -40.7443, -40.7251, -40.8300],
         [-17.3248,  -5.3050, -11.4988,  ..., -41.1802, -41.1236, -41.2818],
         [-18.3950,  -5.8967, -11.9756,  ..., -42.3553, -42.3273, -42.4553]],
        [[-15.7967,  -6.9833, -14.5827,  ..., -41.3168, -41.0326, -40.9567],
         [-18.5133,  -5.7466, -12.1301,  ..., -40.2199, -40.1078, -39.9497],
         [-19.3636,  -5.1246, -12.7134,  ..., -42.1881, -41.9993, -41.9106],
         [-20.3247,  -3.6335, -12.4559,  ..., -40.0285, -39.9912, -39.8106]]])
```
Am I did anything wrong?

**Comment by ArvinZhuang** [comment]
When I add torch.manual_seed(0) at the beginning the outputs will be the same.
does the model from tf pretrained has some randomization when loading the model weights?

**Comment by patrickvonplaten** [comment]
Hmm, there sholud always be the same. Could you add your script to a google colab that allows to run your checkpoint, so that we can debug?

**Comment by ArvinZhuang** [comment]
Hi here is the colab script with my checkpoint uploaded:
https://colab.research.google.com/drive/1Yx7zRkzpaGMMraTkIeIGnJJLnlpUKzQH?usp=sharing

**Comment by patrickvonplaten** [comment]
I think some layers in your T5 model have not been trained and are therefore not saved in your model checkpoint. At initialization this layer is then randomely initialized. 

One thing you can do to verify is to load the model as follows, save it as a PyTorch model and then load the PyTorch model as follows.

```python 
config = T5Config.from_pretrained('t5-base')
model = T5ForConditionalGeneration.from_pretrained(
    "sample_data/model/model.ckpt-1004000", from_tf=True, config=config)
model.save_pretrained("./model")
model = T5ForConditionalGeneration.from_pretrained("/.model")
```

if this command throws a warning that some layers are not initialized, then you know what the problem is.
If not, I will take a look again :-)

**Comment by ArvinZhuang** [comment]
hi, I tried what you said, no warning shows up.
so I guess it is not the case?

Do you think it could be that weights from float64 to float32 or the opposite causing this problem?  because the difference in outputs is tiny. I don't know, this still cannot explain why outputs are different every time.

**Comment by ArvinZhuang** [comment]
any updates?

**Comment by patrickvonplaten** [comment]
Hey @ArvinZhuang, 

I just downloaded the weights and tried to reproduce the error using your code snippet. In my case the output is deterministic, as expected. Could you make sure that you are using the newest version of transformers and try again?

**Comment by ArvinZhuang** [comment]
Hi, the outputs still different on my machine.... very strange.

I'm using transformer v3.3.0
torch v1.6.0
TensorFlow v2.3.1 

![image](https://user-images.githubusercontent.com/46237844/94499979-607f0900-0241-11eb-8b64-3b91d4ef1d31.png)

**Comment by ednussi** [comment]
+1 on the issue

I run:

```
from transformers import AutoTokenizer, AutoModelForSeq2SeqLM, AutoConfig

config = AutoConfig.from_pretrained("Vamsi/T5_Paraphrase_Paws", output_hidden_states=True)  
tokenizer = AutoTokenizer.from_pretrained("Vamsi/T5_Paraphrase_Paws")  
model = AutoModelForSeq2SeqLM.from_pretrained("Vamsi/T5_Paraphrase_Paws", config=config).to('cuda')

def prediction(documents, query):
    querys = [query] * len(documents)
    encoded_encoder_inputs = tokenizer(documents, padding=True, truncation=True, return_tensors="pt").to('cuda')
    encoded_decoder_inputs = tokenizer(querys, padding=True, truncation=True, return_tensors="pt").to('cuda')
    with torch.no_grad():
        outputs = model(input_ids=encoded_encoder_inputs["input_ids"],
                        labels=encoded_decoder_inputs["input_ids"],
                        attention_mask=encoded_encoder_inputs["attention_mask"])
        batch_logits = outputs[1]
        print(batch_logits)

documents = ['my dog is cute!', "so am I."]
query = "who am I?"
prediction(documents, query)    
```

and got:
```
tensor([[[-24.6745, -11.9357, -14.9241,  ..., -45.9948, -45.8151, -45.8455],
         [-23.2342, -10.9913,  -8.3003,  ..., -41.6439, -41.4256, -41.5661],
         [-25.2330, -12.2352, -13.0547,  ..., -47.6261, -47.4068, -47.5662],
         [-21.9263, -11.0629, -11.9107,  ..., -41.9656, -41.8519, -42.0753],
         [-26.9045,  -4.1973, -10.9553,  ..., -43.4384, -43.3700, -43.4003]],

        [[-23.9796, -10.0495, -12.1603,  ..., -44.2115, -44.2157, -44.1649],
         [-24.5771,  -9.9728,  -9.0679,  ..., -42.0522, -41.9042, -41.9449],
         [-27.3110, -11.3050, -12.3826,  ..., -46.7822, -46.7532, -46.8192],
         [-23.5749,  -7.7868, -11.3786,  ..., -44.6472, -44.6335, -44.6697],
         [-32.4967,  -2.0641, -11.7173,  ..., -49.9682, -50.0462, -49.9918]]],
       device='cuda:0')
```

**Comment by ArvinZhuang** [comment]
Hi ednussi, 
yes, rerunning prediction(documents, query) will give deterministic results. 
However, my issue is rerunning the above chunk of code twice (reload everything, including model.). the outputs are different.
Does this also happen to you?

**Comment by ednussi** [comment]
Hi @ArvinZhuang,
Yes, similar to you when I run the model call with same input_ids but different labels I got different outcomes (using pytorch). Realized I miss a part of the print so Edited my comment above to match the output.
Hoping @patrickvonplaten or someone from the team can take a second look.

**Comment by patrickvonplaten** [comment]
Hey @ednussi, if you look into the code, you can see that T5 uses the `labels` to create the `decoder_input_ids` which necessarily do affect the outputs. You can try using deterministic `input_ids` and `decoder_input_ids` and no labels and see if the output stays deterministic (it should).

**Comment by ednussi** [comment]
Thanks for responding @patrickvonplaten, only now saw you got back to me.
Following on your suggestion, I was able to find my error.

To locate my misunderstanding I run:
```
from transformers import AutoTokenizer, AutoModelForSeq2SeqLM, AutoConfig
tokenizer = AutoTokenizer.from_pretrained('t5-base')  
model = AutoModelForSeq2SeqLM.from_pretrained('t5-base').to('cuda')

def get_sent_embed(input_sents, labels):
    encoded_decoder_inputs = tokenizer(labels, padding=True, truncation=True, return_tensors="pt").to('cuda')
    encoded_encoder_inputs = tokenizer(input_sents, padding=True, truncation=True, return_tensors="pt").to('cuda')
    with torch.no_grad():
        outputs = model(input_ids=encoded_encoder_inputs["input_ids"],
                        labels=encoded_decoder_inputs["input_ids"],
                        attention_mask=encoded_encoder_inputs["attention_mask"],
                        return_dict=True)
        logits = outputs.logits[0]
        return logits
    
logits = get_sent_embed(['Test'], ['a'])
logits2 = get_sent_embed(['Test'], ['b'])
print(sum(logits[0] - logits2[0]))

logits = get_sent_embed(['Test', 'Test'], ['a', 'b'])
print(sum(logits[0] - logits[1]))

logits2 = get_sent_embed(['Test', 'Test'], ['aa', 'bb'])
print(all((logits[0] - logits2[0])< 10**-4))
print(all((logits[1] - logits2[1])< 10**-4))
```
which yields:
```
tensor(0., device='cuda:0')
tensor(55281.6445, device='cuda:0')
True
True
```
After looking into the definition of input_ids: `(torch.LongTensor of shape (batch_size, sequence_length), optional) – The sequence used as a prompt for the generation. If None the method initializes it as an empty torch.LongTensor of shape (1,).`

I realized this methods actually runs, as stated in the documentation, over the sequence of inputs, hence the different results from inferencing on ['Test'] and ['Test', 'Test']. Same sequence with different labels do return the same inference results, as expected.
Note: Regarding the last precision print in the end - I saw differences at the 5th floating point for the logits, but that's just floating point precision issues.

@ArvinZhuang hope this helps you as well.

**Comment by ednussi** [comment]
Thanks @patrickvonplaten.
Was following your suggestion, and after reading through the documentation and code of how the `decoder_input_ids` is used, it became clear why it affects the logits and helped clear my confusion.

**Comment by vikas95** [comment]
@ednussi - wanted to ask for your help if you could provide a brief explanation from your reading? Thanks.
